### PR TITLE
Uwp

### PR DESCRIPTION
**Pull request by tianchiTJ**
1. "Particle"
① Add "update_porosity"
② Add "compute_pore_pressure”
③ Add "map_drag_force_coefficient“
④ Add "map_mixture_internal_force”

2."Node"
① Add "update_mixture_internal_force"
② Add "update_drag_force_coefficient"
③ Add function to return "mixture_internal_force"
④ Add function to return "drag_force_coefficient“
⑤ Add "compute_acceleration_velocity_two_phase"

3."Cell"
① Add "compute_nodal_mixture_internal_force"
② Add "compute_nodal_drag_force_coefficient"

4."logger"
Add logger of  "mpm_explicit_two_phase"

### CODE REVIEW

File: include/particle.tcc [hunk 1]
```diff
@@ -386,7 +386,14 @@ bool mpm::Particle<Tdim, Tnphases>::update_volume(unsigned phase, double dt) {
     if (cell_ != nullptr && volume_ != std::numeric_limits<double>::max()) {
 
       Eigen::VectorXd strain_rate = cell_->compute_strain_rate(bmatrix_, phase);
```
Comment (srhgk2) [inline include/particle.tcc]: Use this function to update the volume only.
Define a new function to update the porosity and use the solid mass balance equation.

**Review by srhgk2 — CHANGES_REQUESTED** [review]

File: include/particle.tcc [hunk 2]
```diff
@@ -386,7 +386,12 @@ bool mpm::Particle<Tdim, Tnphases>::update_volume(unsigned phase, double dt) {
...
-      this->volume_ *= (1. + dt * strain_rate.head(Tdim).sum());
+      // Compute incremental of particle volume
+      double dvolume = this->volume_ * dt * strain_rate.head(Tdim).sum();
+      // Compute particle volume of pore fluid
+      double volume_fluid = this->volume_ * this->porosity_ + dvolume;
```
Comment (srhgk2) [inline include/particle.tcc]: No need for the `volume_fluid`

File: include/particle.tcc [hunk 3]
```diff
@@ -398,6 +403,29 @@ bool mpm::Particle<Tdim, Tnphases>::update_volume(unsigned phase, double dt) {
...
+template <unsigned Tdim, unsigned Tnphases>
+bool mpm::Particle<Tdim, Tnphases>::update_porosity(unsigned phase, double dt) {
+  bool status = true;
+  try {
+    // Check if particle has a valid cell ptr and a valid volume
+    if (cell_ != nullptr && volume_ != std::numeric_limits<double>::max()) {
+
+      Eigen::VectorXd strain_rate = cell_->compute_strain_rate(bmatrix_, phase);
+      // Update particle volume
+      this->porosity_ +=
```
Comment (srhgk2) [inline include/particle.tcc]: Correct the porosity update equation.

**Review by srhgk2 — CHANGES_REQUESTED** [review]

File: include/particle.tcc [hunk 4]
```diff
@@ -538,9 +642,8 @@ bool mpm::Particle<Tdim, Tnphases>::map_internal_force(unsigned phase) {
     if (material_.at(phase) != nullptr) {
       // Compute nodal internal forces
       // -pstress * volume
-      cell_->compute_nodal_internal_force(
-          this->bmatrix_, phase, this->volume_ * this->volume_fraction_(phase),
-          -1. * this->stress_.col(phase));
+      cell_->compute_nodal_internal_force(this->bmatrix_, phase, this->volume_,
```
Comment (srhgk2) [inline include/particle.tcc]: Revert to the original function.

**Review by srhgk2 — CHANGES_REQUESTED** [review]

File: include/particle.tcc [hunk 5]
```diff
@@ -672,6 +801,94 @@ bool mpm::Particle<Tdim, Tnphases>::compute_updated_position_velocity(
   return status;
 }
 
+// Compute updated position of the particle
+template <unsigned Tdim, unsigned Tnphases>
+bool mpm::Particle<Tdim, Tnphases>::compute_updated_position_two_phase(
```
Comment (srhgk2) [inline include/particle.tcc]: Let's have two functions
1. `update_velocity`
2. `update_position`

File: include/particle.tcc [hunk 6]
```diff
@@ -672,6 +801,94 @@ bool mpm::Particle<Tdim, Tnphases>::compute_updated_position_velocity(
...
+  } catch (std::exception& exception) {
+    console_->error("{} #{}: {}\n", __FILE__, __LINE__, exception.what());
+    status = false;
+  }
+  return status;
+}
+
+// Compute updated position of the particle based on nodal velocity
+template <unsigned Tdim, unsigned Tnphases>
+bool mpm::Particle<Tdim, Tnphases>::compute_updated_position_velocity_two_phase(
```
Comment (srhgk2) [inline include/particle.tcc]: Define a new function for `update_velocity` using nodal velocity

**Review by srhgk2 — CHANGES_REQUESTED** [review]

File: include/particle.tcc [hunk 7]
```diff
@@ -517,6 +540,45 @@ bool mpm::Particle<Tdim, Tnphases>::compute_stress(unsigned phase) {
...
+      double K = material_.at(pore_fluid)->property("bulk_modulus");
+      // Compute strain rate of the particle
+      // Solid skeleton
+      Eigen::VectorXd strain_rate_solid =
+          cell_->compute_strain_rate(bmatrix_, solid_skeleton);
+      // Pore fluid
+      Eigen::VectorXd strain_rate_fluid =
+          cell_->compute_strain_rate(bmatrix_, pore_fluid);
+      // Calculate pore pressure
+      double dpressure = K / porosity_ *
```
Comment (srhgk2) [inline include/particle.tcc]: remove solid volume fraction.

**Review by srhgk2 — CHANGES_REQUESTED** [review]

File: include/particle.tcc (hunk 7 above)
Comment (srhgk2) [reply include/particle.tcc]: missing (-) sign

**Review by srhgk2 — CHANGES_REQUESTED** [review]

File: include/particle.tcc (hunk 7 above)
Comment (tianchiTJ) [reply include/particle.tcc]: > remove solid volume fraction.

Now the velocities stored are the real velocities of solid and fluid, so I think we need the volume fraction.

**Pull request merged**